### DISCUSSION TITLE
Command upsertion optimization

### DISCUSSION THREAD

**Discussion by Chew** [Chew/JDA-Chewtils#8] [Ideas]
Currently, the bot upserts every command on boot, regardless of what's currently there.

Obviously, this is not ideal for upserting a ton of commands.

My current plan is as follows:

- Deprecate SlashCommand#guildId. This shouldn't be used and should instead be replaced with CommandClient#forceGuildOnly()
- Based on forceGuildOnly() status, retrieve commands either globally or from the selected Guild. When upserting commands, only upsert if it's changed (name, description, removed, added, etc). Because of how slash commands work, this should have no breaking changes. In theory it could be replaced with .updateCommands() I guess?
- Make SlashCommand#guildOnly false by default (because commands are global, it makes no sense for them to be tied to a guild). This may be a breaking change so I may not end up doing this.

In theory, this will only end up upserting _changed_ commands, making boots significantly quicker and testing much easier. 

Feel free to leave thoughts below before I start making it!

**Comment by Andre601** [comment]
So if I understand it correctly would Chewtils pull the commands Discord currently has for the bot and compares them with the ones itself has to then decide what to do?
Or how would the logic behind this look like?

**Reply by Chew** [reply]
Most likely, though it may not be necessary.
With Chewbotcca I just upsert all the commands from the library on boot and let it update manually, which is much quicker.
https://github.com/Chewbotcca/Discord/blob/main/src/main/java/pw/chew/chewbotcca/listeners/ReadyListener.java#L44-L55
This is probably the logic I'd use, somewhat.

**Reply by yHSJ** [reply]
Hi I took a look at the code you use and just have a quick question -- If you update or remove a slash command, does it update or remove the existing command (given the hour for global commands) from the list of commands in discord? I am having a strange "issue" where when I update an existing command or remove a command it creates a new command and leaves the old one. When trying to run the command it doesn't work, but it doesn't remove itself from the list of possible commands in discord.

Is there already a solution to this?

**Reply by Chew** [reply]
Not automatically, you'd either need to restart or, if you use `2.0`, call `CommandClient#upsertInteractions`

As for it leaving the old one, I think I've experienced this, but I think it's a Discord bug tbh.
